cajita/src/Cajita_Parallel.hpp

```diff
@@ -157,6 +157,231 @@ grid_parallel_for( const std::string& label, const ExecutionSpace& exec_space,
     grid_parallel_for( label, exec_space, index_space, work_tag, functor );
 }
 
+//---------------------------------------------------------------------------//
+/*!
+  \brief Execute a functor in parallel with a linear execution
+  policy specified by the set of given index spaces. 4D specialization.
+
+  \tparam FunctorType The functor type to execute.
+
+  \tparam ExecutionSpace The execution space type.
+
+  \tparam NumSpace The number of index spaces.
+
+  \param label Parallel region label.
+
+  \param exec_space An execution space instance.
+
+  \param index_spaces The set of index spaces over which to loop.
+
+  \param functor The functor to execute.
+ */
+template <class FunctorType, class ExecutionSpace, std::size_t NumSpace>
+inline void
+grid_parallel_for( const std::string& label, const ExecutionSpace& exec_space,
+                   const Kokkos::Array<IndexSpace<4>, NumSpace>& index_spaces,
+                   const FunctorType& functor )
+{
+    // Compute the total number of threads needed and the index space offsets
+    // via inclusive scan.
+    int size = index_spaces[0].size();
+    Kokkos::Array<int, NumSpace> exclusive_offsets;
+    Kokkos::Array<int, NumSpace> inclusive_offsets;
+    exclusive_offsets[0] = 0;
+    inclusive_offsets[0] = size;
+    for ( std::size_t sp = 1; sp < NumSpace; ++sp )
+    {
+        size += index_spaces[sp].size();
+        exclusive_offsets[sp] =
+            exclusive_offsets[sp - 1] + index_spaces[sp - 1].size();
+        inclusive_offsets[sp] =
+            inclusive_offsets[sp - 1] + index_spaces[sp].size();
+    }
+
+    // Unroll the index spaces into a linear parallel for.
+    Kokkos::parallel_for(
+        label, Kokkos::RangePolicy<ExecutionSpace>( exec_space, 0, size ),
```

Comment:
Good question. I think this fused version will only be used in cases where launch latency is a dominating factor hence kernel performance is secondary to eliminating latency. With that said, it would be nice to have something that demonstrates this explicitly through a performance test. I have done some similar things in comm where I showed performance of comm on individual slices vs. doing the entire AoSoA at once.